### ISSUE TITLE
MidiDevice.Info all accessible fields are "null".

### ISSUE THREAD

**Issue by Pansmile** [DerekCook/CoreMidi4J#12]
I can't get any not-null MidiDevice.Info from CoreMidi MidiDevices implementations. 
I mean that i can get an Info by
 `MidiDevice.Info info = device.getDeviceInfo();`
 but calling to
 `info.toString()` or `info.getName()` or `info.getDescription()` returns "null".
regular implementations of similar devices always return correct Info.
I'm learning java for three month and the last month i was  working at my first project called MIDI Tell-Match. This is some kind of simple converter of MIDI messages. I'm using CoreMidi4J to be able to send SysEx messages etc.
Anyway, thank you for great job.

**Comment by DerekCook** [comment]
Hi,

Bit busy right now (sigh). Can you provide some sample code to allow us to try and replicate your issue?

**Comment by Pansmile** [comment]
@DerekCook 
Wow, you responded so fast!
here is the simple test-class:

```
import javax.sound.midi.MidiDevice;
import javax.sound.midi.MidiSystem;
import javax.sound.midi.MidiUnavailableException;

public class CoreMidiDeviceInfoTest {
    public static void main(String[] args) throws MidiUnavailableException {
        //Getting all devices info from MidiSystem.
        MidiDevice.Info[] infoArray = MidiSystem.getMidiDeviceInfo();
        for (MidiDevice.Info anInfo : infoArray) {
            MidiDevice device = MidiSystem.getMidiDevice(anInfo);
            //Checking if device is a CoreMidi implementation.
            if (device.toString().contains("CoreMidi")) {
                System.out.println(anInfo);
                System.out.println(anInfo.getName());
                System.out.println(anInfo.getDescription());
                System.out.println();
            }
        }
    }
}

```

On my system it prints to the console this text:

> null
> null
> null

null
null
null

null
null
null

null
null
null

**Comment by brunchboy** [comment]
Strange. On my system I get:

```
> java CoreMidiDeviceInfoTest
CoreMIDI4J - Bus 1
CoreMIDI4J - Bus 1
IAC Driver

CoreMIDI4J - Network
CoreMIDI4J - Network


CoreMIDI4J - Bus 1
CoreMIDI4J - Bus 1
IAC Driver

CoreMIDI4J - Network
CoreMIDI4J - Network

```

**Comment by Pansmile** [comment]
@brunchboy 
Oops. Seems to be depended on my configuration. 
I'm just wondering what I'm doing wrong. Hm.

**Comment by brunchboy** [comment]
I am not sure, it seems like there should be nothing you could do that would cause that, but I am completely buried preparing for an all-night light show on Saturday, and don‘t know when I might have time to help investigate.

**Comment by Pansmile** [comment]
@brunchboy 
Oh, I understand your feelings, i'm working in a theatre of opera and ballet :).
Hope some day someone will help me.

**Comment by DerekCook** [comment]
Hi, I might have been quick of the blocks, but James got there before me ;-)

I found the time tonight to run your code and this is what  get on my system (Edirol UA-25 and Mobile Keys 25 USB keyboard)

> CoreMIDI4J - UA-25
> CoreMIDI4J - UA-25
> UA-25
> 
> CoreMIDI4J - Mobile Keys 25
> CoreMIDI4J - Mobile Keys 25
> Mobile Keys 25
> 
> CoreMIDI4J - UA-25
> CoreMIDI4J - UA-25
> UA-25
> 
> CoreMIDI4J - Mobile Keys 25
> CoreMIDI4J - Mobile Keys 25
> Mobile Keys 25

So, not sure what is happening on your system. BTW you get two copies of a device with your code, as one is a source, and one is a destination, but your code does not differentiate between sources and destinations. Just thought I'd explain that one :-)

I'd try putting in to your code the following.

`System.out.println(device.toString());`

Before printing out the other stuff, just to see if you are getting the expected device name.

In your println code for dumping parameters, I also added 

`System.out.println(anInfo.getVendor());` to ensure that was as expected (it was)

I'd also put a break point on the line 

`MidiDevice device = MidiSystem.getMidiDevice(anInfo);`

And check the values in `anInfo` to see if it gives you any clues. 

I'd also check what **Audio MIDI Setup** (found under _Applications/Utilities_) shows for your system if those values are empty, it might explain the issue. If nothing else, it would be good to rule out

HTH

Cheers
Derek

You know, that's the first bit of messing with code that I have done for a month! :-) Too many things to do, to little time!

**Comment by Pansmile** [comment]
@DerekCook 
Hi!
I'm really glad that you took a look at my code.
Of course, I understand that my example is putting sources and destinations in one pile :).
I've made some changes to my test class to make it more clear:

```
public class CoreMidiDeviceInfoTest {
    public static void main(String[] args) throws MidiUnavailableException {
        //Getting all devices info from MidiSystem.
        MidiDevice.Info[] infoArray = MidiSystem.getMidiDeviceInfo();
        System.out.println("Sources:");
        for (MidiDevice.Info anInfo : infoArray) {
            MidiDevice device = MidiSystem.getMidiDevice(anInfo);
            //Filtering destinations.
            if (device.toString().contains("MidiSource") || device.toString().contains("MidiOut")) {
                printDevice(device, anInfo);
            }
        }
        System.out.println();
        System.out.println("Destinations:");
        for (MidiDevice.Info anInfo : infoArray) {
            MidiDevice device = MidiSystem.getMidiDevice(anInfo);
            //Filtering destinations.
            if (device.toString().contains("MidiSource") || device.toString().contains("MidiIn")) {
                printDevice(device, anInfo);
            }
        }
    }

    public static void printDevice(MidiDevice device, MidiDevice.Info anInfo) {
        System.out.println(device.toString());
        System.out.println(anInfo);
        System.out.println(anInfo.getName() + " " +  anInfo.getDescription() + " " + anInfo.getVendor() + "\n");
    }
}
```

Here wat I've got from the console:

> Sources:
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@7e6cbb7a
> null
> null null Unknown vendor
> 
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@76fb509a
> null
> null null Unknown vendor
> 
> com.sun.media.sound.MidiOutDevice@41cf53f9
> CTRL
> CTRL nanoKEY2 CTRL KORG INC.
> 
> com.sun.media.sound.MidiOutDevice@68de145
> Launchpad Mini 11
> Launchpad Mini 11 Launchpad Mini Launchpad Mini 11 Focusrite A.E. Ltd
> 
> com.sun.media.sound.MidiOutDevice@5ce65a89
> MIDI Monitor (Untitled)
> MIDI Monitor (Untitled) MIDI Monitor (Untitled) Snoize
> 
> com.sun.media.sound.MidiOutDevice@69d0a921
> QLab
> QLab QLab Snoize
> 
> Destinations:
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@7e6cbb7a
> null
> null null Unknown vendor
> 
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@76fb509a
> null
> null null Unknown vendor
> 
> com.sun.media.sound.MidiInDevice@728938a9
> KEYBOARD
> KEYBOARD nanoKEY2 KEYBOARD KORG INC.
> 
> com.sun.media.sound.MidiInDevice@533ddba
> Launchpad Mini 11
> Launchpad Mini 11 Launchpad Mini Launchpad Mini 11 Focusrite A.E. Ltd

As you can see, regular implementations returns proper values of Info.

Audio Midi Setup shows proper information for visible devices:
(Sorry for Russian language, but i thing everything is clear anyway).
![2016-03-03 12 04 20](https://cloud.githubusercontent.com/assets/17011774/13489454/337eb8c6-e138-11e5-8054-de54705e7159.png)

I added a breakpoint after `MidiDevice.Info[] infoArray = MidiSystem.getMidiDeviceInfo();`
to watch all devices info.
Here is what I've got:

![2016-03-03 12 08 29](https://cloud.githubusercontent.com/assets/17011774/13489634/33034e92-e139-11e5-92b7-813697138f2d.png)
![2016-03-03 12 10 15](https://cloud.githubusercontent.com/assets/17011774/13489635/333070a2-e139-11e5-8607-21a2db89dfa9.png)

It seems like only version of device is captured properly in CoreMidi implementations.

P.S. Sorry for my English.

**Comment by DerekCook** [comment]
Hi,

Please don't apologise for your English. How do you think my Russian will compare with how well you are writing English?! :-)

Running your latest example, this is what I get

> Sources:
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@1b1c7295
> CoreMIDI4J - UA-25
> CoreMIDI4J - UA-25 UA-25 EDIROL
> 
> com.sun.media.sound.MidiOutDevice@45ee5b36
> UA-25
> UA-25 UA-25 EDIROL
> 
> Destinations:
> uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource@1b1c7295
> CoreMIDI4J - UA-25
> CoreMIDI4J - UA-25 UA-25 EDIROL
> 
> com.sun.media.sound.MidiInDevice@30f02a6d
> UA-25
> UA-25 UA-25 EDIROL

Looking at the code base, a CoreMidiDeviceInfo is only created in one place CoreMidiDeviceProvider.cpp, and I am not sure why the code is succeeding on our systems but not yours. We may need to add some debug code in to trace what is happening. Also looking at the code, I am not checking the return results from the CoreMIDI calls to get the properties from the end point, and perhaps I should do, which might be highlight the cause of the error. 

Can you confirm that you are using the latest version of the software and verify that you have both CoreMIDI4J.jar and libCoreMidi4J.dylib installed in your /Library/Java/Extensions folder? I am expecting that all this is fine, but it's nice to eliminate potential causes of problems.

Also, what version of Java are you using?

**Comment by DerekCook** [comment]
Also, can you try running the following code fragment and report back what it provides?

```
MidiDevice.Info devices[] = MidiSystem.getMidiDeviceInfo();

// Test to see what devices we can pickup from the system
System.out.println("Number of Devices: " + devices.length);

for (int i = 0; i < devices.length; i += 1 ) {

    MidiDevice device = MidiSystem.getMidiDevice(devices[i]);

    System.out.println("  Device " + i + " " + devices[i].getName() + " " + device.getMaxReceivers() + " " + device.getMaxTransmitters());

}
```

**Comment by Pansmile** [comment]
Hi,
I'm using java 8 update 74 and jdk1.8.0_74, version of CoreMidi4j is 0.4, both jar and dylib are installed in  
/Library/Java/Extensions.

Running your code fragment provides this:

> Number of Devices: 14
>   Device 0 null 0 -1
>   Device 1 null 0 -1
>   Device 2 null -1 0
>   Device 3 null -1 0
>   Device 4 null -1 0
>   Device 5 null -1 0
>   Device 6 Gervill -1 0
>   Device 7 KEYBOARD 0 -1
>   Device 8 Launchpad Mini 11 0 -1
>   Device 9 CTRL -1 0
>   Device 10 Launchpad Mini 11 -1 0
>   Device 11 QLab -1 0
>   Device 12 MIDI Monitor (Untitled) -1 0
>   Device 13 Real Time Sequencer -1 -1

P.S. A small and bad quality photo of Mariinsky Ballet performing Swan Lake for you right now :)
![picc](https://cloud.githubusercontent.com/assets/17011774/13505799/e5c8ac96-e18b-11e5-8d4a-926ebfa4b16e.png)

**Comment by DerekCook** [comment]
Love the picture. 

OK, all of that means we are going to need to dig deeper. The fact that that sample code I provided is giving the same NULL results rules it out as being a quirk of your code. Again, this is what I expected, but wanted to rule it out. :-)

Over the weekend (hopefully) I will create a new branch of the code and add more error checking and debug code into the native code that creates the CoreMidiDeviceInfo Object.

Please bear with me in looking at this. This is a small hobby at a time when I am very busy with work.

Cheers
Derek

**Comment by brunchboy** [comment]
Thanks for looking at this when you can Derek. I would try to jump in and help but I, too, am very busy with work _and_ preparing for the all-night light and laser show I will be running Saturday to Sunday. Which will be using CoreMidi4J heavily. Thankfully it has been working a treat on my system! And @Pansmile thank you for the lovely photo! I hope to take some (and some video) of the show Saturday, it will be the first time I use my [Afterglow](https://github.com/brunchboy/afterglow#afterglow) with our full lighting rig (at least if we can avoid having any more moving head lights die before the weekend is over).

**Comment by Pansmile** [comment]
@DerekCook 
Hope we will find out where the problem is hiding. 
Anyway, I can wait as long as I had to. I needed proper mididevice.Info for  automatically detecting sources or destination of my midi converters after reconnecting them. For now I added a possibility to reconnect them manually.
Hope that I'll finish first version of my software to upload it to the gig hub.
I'm not sure if it would be useful for someone, cause it's very specific:).
@brunchboy 
Can't wait to see your show pics. Good luck with it.

**Comment by DerekCook** [comment]
Hi, I haven't figured out a neater way of doing this (GIT will not let me attach either the dylib or a zip to this message, despite claiming to allow ZIPs to be attached), and I am off out for the day now. Anyhow, there is a branch called Issue12Investigation. If you download the branch zip file and extract the contents, under the **CoreMidi4J/Build/Release** folder, you will find a new dylib which is giving some debug info when it creates a CoreMidiDeviceInfo object. So copy that dylib to your *\* /library/java/extensions*\* folder and run it.

This debug build provides the following debug info on my system (for UA25 source and destination end points. 

> > *\* Debug - End Point Ref: 179413011
> >  *\* Debug - Name         : UA-25
> >  *\* Debug - Description  : UA-25
> >  *\* Debug - Manufacturer : EDIROL
> >  *\* Debug - Version      : 0
> >  *\* Debug - UID          : -536838920
> >  *\* Debug - End Point Ref: 179413012
> >  *\* Debug - Name         : UA-25
> >  *\* Debug - Description  : UA-25
> >  *\* Debug - Manufacturer : EDIROL
> >  *\* Debug - Version      : 0
> >  *\* Debug - UID          : 219594876

This is the property info being requested in the native code before a DeviceInfo object is built.

So it will be interesting to see what it shows on yours. As well as the obvious wondering how the strings are coming back at this point, I am interested in how the end point reference looks like. It is the Core MIDI reference for the end point, and I am wondering if it is 0 or -1 or some other error value.

**Comment by Pansmile** [comment]
Hi,
Thanks. Where should I look for this debug info? Nothing new in std err and std out.

**Comment by DerekCook** [comment]
Apologies. In my haste to get out of the house yesterday, I forgot to push the commit. If you try again with the download from the Issue12Investigation branch, then hopefully that will work now!

The debug messages are written to stdout and should appear in the same console as your Java console messages.

**Comment by DerekCook** [comment]
Just did a quick test myself of trying to print before the pointers are initialised,and I get a fatal error, so I'll make the code more resilient and republish.

**Comment by DerekCook** [comment]
Have improved the code to print out the OSStatus codes as well.

**Comment by Pansmile** [comment]
@DerekCook 
Thanks!
Now there is no dylib. should i replace the jar?

**Comment by Pansmile** [comment]
Found the dylib under archive.

**Comment by Pansmile** [comment]
@DerekCook 
I've got fatal error.

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff8b2fcf92, pid=1018, tid=4867
#
# JRE version: Java(TM) SE Runtime Environment (8.0_74-b02) (build 1.8.0_74-b02)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.74-b02 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libsystem_c.dylib+0xf92]  strlen+0x12
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Volumes/Files/TestSoftware/MidiTallMatch/hs_err_pid1018.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
 ** Debug - End Point Ref: 1814557
 ** Debug - Name         : 
Process finished with exit code 134

```

**Comment by DerekCook** [comment]
I don't know why the latest dylib was not included, but have just rebuilt and pushed the commit. I have downloaded and checked. Please try the one in the **CoreMidi4J/Build/Release** folder

That crash was me messing about to see what happened if there was a null pointer (and you can see the result!)

**Comment by Pansmile** [comment]
@DerekCook 
well, I've got this:

```
Number of Devices: 14
  Device 0 null 0 -1
  Device 1 null 0 -1
  Device 2 null -1 0
  Device 3 null -1 0
  Device 4 null -1 0
  Device 5 null -1 0
  Device 6 Gervill -1 0
  Device 7 KEYBOARD 0 -1
  Device 8 Launchpad Mini 11 0 -1
  Device 9 CTRL -1 0
  Device 10 Launchpad Mini 11 -1 0
  Device 11 MIDI Monitor (Untitled) -1 0
  Device 12 QLab -1 0
  Device 13 Real Time Sequencer -1 -1
 ** Debug - End Point Ref: 00000000, 001bb015
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, cecf0dc4
 ** Debug - End Point Ref: 00000000, 001bb019
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 13d8e0df
 ** Debug - End Point Ref: 00000000, 001bb016
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 16938d10
 ** Debug - End Point Ref: 00000000, 001bb01a
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 659e4a47
 ** Debug - End Point Ref: 00000000, 001bb03b
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 34506d8f
 ** Debug - End Point Ref: 00000000, 001bb03f
 ** Debug - Name         : 00000000, NULL Pointer
 ** Debug - Description  : 00000000, NULL Pointer
 ** Debug - Manufacturer : 00000000, NULL Pointer
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 4b4d5ea3
```

**Comment by DerekCook** [comment]
OK. Thanks. We got there! :-)

So, we have a non zero end point reference, which is good. But we are getting null pointers on the device names when building the C strings, however the OS Status code (the first number) which is returned from `CFStringGetCStringPtr` is not indicating an error, which is strange. The documentation for this function does warn you `/* May return NULL at any time; be prepared for NULL */`, so the question is under what circumstances can you expect NULL?

I wonder if it's the fact that I am using `kCFStringEncodingMacRoman` in the following snippet, and your language encoding is something else?

const`char *debugName = CFStringGetCStringPtr ( name, kCFStringEncodingMacRoman );`

That's all I have time for today. Will need to have a think on what to try next.

**Comment by Pansmile** [comment]
@DerekCook 
Hah!
It seems that we found weak place. I switched the system language to English and here what I've got:

```
Number of Devices: 12
  Device 0 CoreMIDI4J - KEYBOARD 0 -1
  Device 1 CoreMIDI4J - Launchpad Mini 11 0 -1
  Device 2 CoreMIDI4J - CTRL -1 0
  Device 3 CoreMIDI4J - Launchpad Mini 11 -1 0
  Device 4 CoreMIDI4J - MIDI Monitor (Untitled) -1 0
  Device 5 Gervill -1 0
  Device 6 KEYBOARD 0 -1
  Device 7 Launchpad Mini 11 0 -1
  Device 8 CTRL -1 0
  Device 9 Launchpad Mini 11 -1 0
  Device 10 MIDI Monitor (Untitled) -1 0
  Device 11 Real Time Sequencer -1 -1
 ** Debug - End Point Ref: 00000000, 001ba015
 ** Debug - Name         : 00000000, KEYBOARD
 ** Debug - Description  : 00000000, nanoKEY2
 ** Debug - Manufacturer : 00000000, KORG INC.
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, cecf0dc4
 ** Debug - End Point Ref: 00000000, 001ba019
 ** Debug - Name         : 00000000, Launchpad Mini 11
 ** Debug - Description  : 00000000, Launchpad Mini 11
 ** Debug - Manufacturer : 00000000, Focusrite A.E. Ltd
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 13d8e0df
 ** Debug - End Point Ref: 00000000, 001ba016
 ** Debug - Name         : 00000000, CTRL
 ** Debug - Description  : 00000000, nanoKEY2
 ** Debug - Manufacturer : 00000000, KORG INC.
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 16938d10
 ** Debug - End Point Ref: 00000000, 001ba01a
 ** Debug - Name         : 00000000, Launchpad Mini 11
 ** Debug - Description  : 00000000, Launchpad Mini 11
 ** Debug - Manufacturer : 00000000, Focusrite A.E. Ltd
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 659e4a47
 ** Debug - End Point Ref: 00000000, 001ba03c
 ** Debug - Name         : 00000000, MIDI Monitor (Untitled)
 ** Debug - Description  : 00000000, MIDI Monitor
 ** Debug - Manufacturer : 00000000, Snoize
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 34506d8f
```

**Comment by DerekCook** [comment]
Brilliant! At least we know what it is now. Will need to read up on, and investigate different character encodings to find the best one to use. Thanks for that.

**Comment by Pansmile** [comment]
@DerekCook 
I'l try to check some other system languages to find out if this is not only Russian encoding problem.

**Comment by Pansmile** [comment]
@DerekCook 
Well, I checked few languages and here is some statistics:
Everything works well with:
English, French, German, Chinese, Italiano, Spanish, Dutch, Korean, Japanese, Danish, Finish, Norwegian, Swedish, Turkish, Arabic, Thai, Croatian, Hebrew, Romanian.
Null returns with:
Greek, Polish, Czech, Hungarian, Slovak, Ukrainian, Russian, Belarusian,.

Did not try enything  else cause I have to go now. Hope this will help somehow:).
Have a nice day.

**Comment by DerekCook** [comment]
You were pretty busy in these checks

So, is it a Cyrillic alphabet issue? What I will do next when I get five minutes is some more debug code that will attempt to print out in different encodings.

**Comment by Pansmile** [comment]
I don’t think so. Maybe it's about alphabets with irregular symbols. Cause Greek Use it's own alphabet, and Polish, Czech, Hungarian and Slovak use Latin alphabet with some additional symbols. but on the other hand, Hebrew use it's own alphabet and everything works well as well as on the Arabic. and Finish, Dutch, Norwegian and even French use irregular symbols and everything was OK. 
The only thing I figured out is that Eastern Europe and maybe Slavic languages  are in the risk group. :)

**Comment by DerekCook** [comment]
OK. Just curious in case it gives a pointer as to what encoding might work for all cases. Or possibly, I might have to try different encodings as each returns NULL, try the next one. We'll see. Will probably look at this again tomorrow night.

**Comment by Pansmile** [comment]
Ok:).

**Comment by DerekCook** [comment]
Hi. Sorry, It's been one heck of a week in work!. So little time for this, but I had time tonight to do a quick code change; the latest build is on the Issue12Investigation branch. 

Instead of hard coding the encoding string in CFStringGetCStringPtr, I found a system call called CFStringGetSystemEncoding(), which will hopefully return whatever encoding you are using. Whether or not it results in a non-null string we need to check. It works on my system, but can you check yours? Let me know.

**Comment by Pansmile** [comment]
@DerekCook 
Hi.
Everything works as charm now!
It's time to delete a lot of code that I used for managing this null-situation :).

**Comment by DerekCook** [comment]
That's fantastic. Sometime in the next few days I will make the change to the main branch and do a formal release of the fix. I'll post back here when it's done, and if you can do a quick test to verify that would be great.

Thank you for your patience in this.

**Comment by Pansmile** [comment]
@DerekCook 
Ooops.
Unfortunately, it seems like I was too fast in conclusions.
Debug info is now right, but `device.getDeviceInfo.getName` or whatever is still "null".

```
 ** Debug - End Point Ref: 00000000, 001b701d
 ** Debug - Name         : 00000000, MIDI
 ** Debug - Description  : 00000000, Axiom AIR 49
 ** Debug - Manufacturer : 00000000, M-Audio
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 3a74b08d
 ** Debug - End Point Ref: 00000000, 001b7020
 ** Debug - Name         : 00000000, HyperControl
 ** Debug - Description  : 00000000, Axiom AIR 49
 ** Debug - Manufacturer : 00000000, M-Audio
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 2e80811f
 ** Debug - End Point Ref: 00000000, 001b7023
 ** Debug - Name         : 00000000, Extern MIDI
 ** Debug - Description  : 00000000, Axiom AIR 49
 ** Debug - Manufacturer : 00000000, M-Audio
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 429c2fab
 ** Debug - End Point Ref: 00000000, 001b7026
 ** Debug - Name         : 00000000, MIDI
 ** Debug - Description  : 00000000, Pro24DSP
 ** Debug - Manufacturer : 00000000, Focusrite A.E.
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 0d91090b
 ** Debug - End Point Ref: 00000000, 001b701e
 ** Debug - Name         : 00000000, MIDI
 ** Debug - Description  : 00000000, Axiom AIR 49
 ** Debug - Manufacturer : 00000000, M-Audio
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, a4eb8bed
 ** Debug - End Point Ref: 00000000, 001b7021
 ** Debug - Name         : 00000000, HyperControl
 ** Debug - Description  : 00000000, Axiom AIR 49
 ** Debug - Manufacturer : 00000000, M-Audio
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 241fc842
 ** Debug - End Point Ref: 00000000, 001b7027
 ** Debug - Name         : 00000000, MIDI
 ** Debug - Description  : 00000000, Pro24DSP
 ** Debug - Manufacturer : 00000000, Focusrite A.E.
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 4d1582ed
 ** Debug - End Point Ref: 00000000, 001b703f
 ** Debug - Name         : 00000000, MIDI Monitor (Untitled)
 ** Debug - Description  : 00000000, MIDI Monitor
 ** Debug - Manufacturer : 00000000, Snoize
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 34506d8f
 ** Debug - End Point Ref: 00000000, 001b7043
 ** Debug - Name         : 00000000, QLab
 ** Debug - Description  : 00000000, QLab
 ** Debug - Manufacturer : 00000000, Snoize
 ** Debug - Version      : ffffd5ad, 00000000
 ** Debug - UID          : 00000000, 4b4d5ea3
```

When  trying
`MidiDevice.Info[] infoArray = MidiSystem.getMidiDeviceInfo();`

I've got this in debugger:

![2016-03-10 0 19 01](https://cloud.githubusercontent.com/assets/17011774/13650857/c4d259c8-e655-11e5-86cb-ea79cb46b68a.png)

**Comment by DerekCook** [comment]
That's OK, as I was only messing with the debug info, not the strings being used to build the object. I've just done another quick update on the Issue12Investigation branch to use the function call in the code to build the object, so maybe you can give that a go.

**Comment by brunchboy** [comment]
It’s great to see you guys found a solution. I still feel like I am recovering from my all-night show this past weekend; I certainly am still catching up at work. Here’s a photo of when we were testing the setup, with CoreMidi4J driving the Ableton Push. :grinning: 

![neural kinetics setup](https://cloud.githubusercontent.com/assets/2228869/13659968/93f4ee42-e64b-11e5-9910-143ae050ac79.jpg)

And here is a shot from behind my controls during the middle of the event, having added a Novation Launchpad Pro to the mix.

![running nk](https://cloud.githubusercontent.com/assets/2228869/13660003/057a05fc-e64c-11e5-864f-d1dd95683ed8.jpg)

Many more photos are in the event [album](https://www.facebook.com/media/set/?set=a.10208880909284270.1073741854.1360869319&type=3) on Facebook, and there are a couple of videos on the [Deep Symmetry page](https://www.facebook.com/DeepSymmetry/) although I did not have time to shoot any myself!

**Comment by Pansmile** [comment]
@DerekCook 
Many thanks! Now everything works great! I'm so happy that we solved this problem.
@brunchboy 
I just took a look at the pictures and videos. Wow! Looks fantastic!

**Comment by brunchboy** [comment]
Thank you , @Pansmile—it is a much less professional production than Mariinsky Ballet, but our guests had a good time. And thanks for your help in identifying and fixing this system encoding issue!

**Comment by Pansmile** [comment]
@brunchboy 
Maybe less professional but not less complex:).
Anyway, my job is the sound and everything connected with lights seems complex to me. Our light team contain about fourty-fifty people per shift (I don't know actual number). That's why I think it is less complex for them to prepare the rig together.

**Comment by DerekCook** [comment]
Having just one person to help me on lights would be a luxury! 

Good pics.

I'll copy the fix into the main branch and do a formal build over the weekend and advise when that is ready for checking. Once we are all happy then I will close this issue

**Comment by DerekCook** [comment]
Hi,

I haven't put this into the main branch yet, but I have tidied up the code for getting the device info and I have removed the debug code and pushed it onto the Issue12Investigation branch. If you could get the dylib from the Issue12Investigation branch and check it out to ensure that everything is still working, I will merge into the main branch as soon as I can.

**Comment by Pansmile** [comment]
@DerekCook 
Hi,
Sorry for long standby, I was quite busy. Just tried last version of Issue12investigation. Everything seems to be ok.

**Comment by DerekCook** [comment]
Hi

Thanks for the feedback. I will do a formal release of the fix once I am back home on the weekend. Thanks for your support in diagnosing this

**Comment by DerekCook** [comment]
V0.5 now released on the main branch with the fix for this issue. I'll now delete the investigation branch

**Comment by brunchboy** [comment]
That’s great, Derek.

I have discovered something I would like to integrate into CoreMidi4J before we do the 1.0 release. In working with the USB video display I mentioned in email, I see that the implementers of [usb4java](http://usb4java.org) have managed to embed their native libraries inside the `jar` file, so that people can use it as an ordinary Java library, and it can be used as a normal Maven/Leiningen/etc. dependency, with no need for the developer or end user of an application which depends on it to install anything anywhere. I would love to study how they did that and replicate it so that I can take advantage of that kind of no-fuss use within Afterglow.

**Comment by DerekCook** [comment]
```
Hi, 

  Well that's how OSXMIDI4J was meant to work as well, but it always
  had permissions problems, and stuffed the native lib in places you
  would not have expected. So hope these USB4Java people have a
  better solution. Worth trying of course. 

  Cheers
  Derek

  On 19/03/2016 19:19, James Elliott wrote:


  That’s great, Derek.
  I have discovered something I would like to integrate into
    CoreMidi4J before we do the 1.0 release. In working with the USB
    video display I mentioned in email, I see that the implementers
    of usb4java
    have managed to embed their native libraries inside the jar
    file, so that people can use it as an ordinary Java library, and
    it can be used as a normal Maven/Leiningen/etc. dependency, with
    no need for the developer or end user of an application which
    depends on it to install anything anywhere. I would love to
    study how they did that and replicate it so that I can take
    advantage of that kind of no-fuss use within Afterglow.
  —
    You are receiving this because you modified the open/close
    state.
    Reply to this email directly or view
      it on GitHub
```

**Comment by brunchboy** [comment]
They definitely seem to be on to something, as it evidently works seamlessly on all platforms.

We may have an extra challenge because of the fact that we want to add a MIDI SPI, but still it may be able to work fine for a single Java process when using this self-installing approach, and also work as an extension when installed in the extensions directory. It will be worth exploring when I come up for air.

**Comment by DerekCook** [comment]
Yes, by all means. If it works then I am interested.

**Comment by Pansmile** [comment]
@DerekCook , @brunchboy  Hi, everyone. 
I just noticed that on last release and issue12Investigation, i could not get timestamps to work. As i remember, it used to work on the previous release. Does anyone can check it?

**Comment by Pansmile** [comment]
Very strange. I use snoize MIDI-monitor. if i create a message and send it with default timeStamp everything works well, if i set any timeStamp different to -1, Every time i sand a message with this timeStamp, MIDI-monitor shows wrong time of catching it. 
![2016-03-20 18 35 31](https://cloud.githubusercontent.com/assets/17011774/13905228/1fd92420-eed3-11e5-9ee0-853b636eb79a.png)

just look at the time column for time of a SysEx message.
